{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28f2bf65",
   "metadata": {},
   "source": [
    "This notebook should be run every time the PLPs selection process changes.\n",
    "\n",
    "Here we filter samples variants so, that only PLPs are left. This is done to make it easier to collect dataset for the regression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c643b98",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ukbb_recessive.data_collection.variants import VariantFeatures\n",
    "import glob\n",
    "import os\n",
    "import datetime\n",
    "import multiprocessing as mp\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aebbc0e4",
   "metadata": {},
   "source": [
    "# Original PLPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dca6b621",
   "metadata": {},
   "outputs": [],
   "source": [
    "# original files with samples variants generated in RAP per chromosome\n",
    "rap_files = glob.glob(\".../450k/RAP_output_per_chr/*.all_parts_final.csv\")\n",
    "\n",
    "## file, that contains all PLPs found in the cohort\n",
    "all_plps_file = (\n",
    "    \".../450k/plp_selection/basic/new_gene_names/new_freq/new_relatedness/\"\n",
    "    \"all_chr_total_presumable_plps_HFE_final_sorted.txt\"\n",
    ")\n",
    "\n",
    "# output folder for PLPs\n",
    "## bugfix with gnomad frequency -- now in %\n",
    "output_folder= '.../450k/RAP_output_per_chr/filtered_plps/basic/new_gene_names/new_freq/new_relatedness'\n",
    "\n",
    "\n",
    "# create output folder if not exists\n",
    "if not os.path.exists(output_folder):\n",
    "    os.makedirs(output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51001d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter out non-PLP variants for every chromosome\n",
    "\n",
    "VariantFeatures().filter_plps_in_samples(\n",
    "    rap_files = rap_files, \n",
    "    output_folder = output_folder, \n",
    "    all_plps_file = all_plps_file\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec453fac",
   "metadata": {},
   "outputs": [],
   "source": [
    "print (\"Filtered PLPs files:\\n\")\n",
    "\n",
    "! ls -lah $output_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc14eb69",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e9970d49",
   "metadata": {},
   "source": [
    "# Numbers check for the paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d93dba0",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_homozygous=True\n",
    "\n",
    "cohort_plp_files = glob.glob(f\"{output_folder}/chr*\")\n",
    "\n",
    "cohort_plps = VariantFeatures().read_sample_plps(cohort_plp_files, filter_homozygous=filter_homozygous).drop(['raw pos', 'raw ref', 'raw alt', 'alleles', 'GT.alleles'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15e10a4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print (f\"Number of PLPs in the cohort:\", cohort_plps.shape[0])\n",
    "print (f\"Number of unique PLPs:\", cohort_plps[['chrom', 'pos', 'ref', 'alt']].drop_duplicates().shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4588bc66",
   "metadata": {},
   "outputs": [],
   "source": [
    "homozygous = cohort_plps[cohort_plps['GT'] == '1/1']\n",
    "\n",
    "print (f\"Number of hom PLPs in the cohort:\", homozygous.shape[0])\n",
    "print (f\"Number of unique hom PLPs:\", homozygous[['chrom', 'pos', 'ref', 'alt']].drop_duplicates().shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bda05cc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_plps = pd.read_csv(all_plps_file, sep='\\t')\n",
    "\n",
    "all_plps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18f1126e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ukbb_recessive",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
