{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28f2bf65",
   "metadata": {},
   "source": [
    "This notebook should be run every time the PLPs selection process changes.\n",
    "\n",
    "Here we filter samples variants so, that only PLPs are left. This is done to make it easier to collect dataset for the regression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4c643b98",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ukbb_recessive.data_collection.variants import VariantFeatures\n",
    "import glob\n",
    "import os\n",
    "import datetime\n",
    "import multiprocessing as mp\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aebbc0e4",
   "metadata": {},
   "source": [
    "# Original PLPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dca6b621",
   "metadata": {},
   "outputs": [],
   "source": [
    "# original files with samples variants generated in RAP per chromosome\n",
    "rap_files = glob.glob(\".../450k/RAP_output_per_chr/*.all_parts_final.csv\")\n",
    "\n",
    "## file, that contains all PLPs found in the cohort\n",
    "all_plps_file = (\n",
    "    \".../450k/plp_selection/basic/new_gene_names/new_freq/new_relatedness/\"\n",
    "    \"all_chr_total_presumable_plps_HFE_final_sorted.txt\"\n",
    ")\n",
    "\n",
    "# output folder for PLPs\n",
    "## bugfix with gnomad frequency -- now in %\n",
    "output_folder= '.../450k/RAP_output_per_chr/filtered_plps/basic/new_gene_names/new_freq/new_relatedness'\n",
    "\n",
    "\n",
    "# create output folder if not exists\n",
    "if not os.path.exists(output_folder):\n",
    "    os.makedirs(output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "51001d65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Entering `filter_plps_in_samples` function...\n",
      "Number of total PLPs: 54758\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 22/22 [39:46<00:00, 108.48s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Function `filter_plps_in_samples` finished, result written in `.../450k/RAP_output_per_chr/filtered_plps/basic/new_gene_names/new_freq/new_relatedness`.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# filter out non-PLP variants for every chromosome\n",
    "\n",
    "VariantFeatures().filter_plps_in_samples(\n",
    "    rap_files = rap_files, \n",
    "    output_folder = output_folder, \n",
    "    all_plps_file = all_plps_file\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ec453fac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered PLPs files:\n",
      "\n",
      "total 73M\n",
      "drwxrwx--- 2 gelana bioinf  937 Mar  5 11:25 .\n",
      "drwxrwx--- 3 gelana bioinf 1.3K Mar  5 10:46 ..\n",
      "-rwxrwx--- 1 gelana bioinf 1.4M Mar  5 11:23 chr10.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 3.1M Mar  5 11:25 chr11.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 2.0M Mar  5 11:27 chr12.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 1.8M Mar  5 11:11 chr13.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 2.4M Mar  5 11:07 chr14.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 2.3M Mar  5 11:02 chr15.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 3.0M Mar  5 10:57 chr16.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 3.0M Mar  5 10:49 chr17.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 807K Mar  5 11:19 chr18.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 2.0M Mar  5 11:04 chr19.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 5.2M Mar  5 11:18 chr1.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 559K Mar  5 10:52 chr20.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 1.3M Mar  5 11:10 chr21.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 836K Mar  5 10:53 chr22.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 3.6M Mar  5 11:00 chr2.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 2.6M Mar  5 10:51 chr3.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 2.3M Mar  5 11:14 chr4.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 2.3M Mar  5 10:55 chr5.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 9.1M Mar  5 11:21 chr6.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 3.5M Mar  5 11:12 chr7.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 1.7M Mar  5 11:06 chr8.all_parts_final.csv\n",
      "-rwxrwx--- 1 gelana bioinf 3.3M Mar  5 11:09 chr9.all_parts_final.csv\n"
     ]
    }
   ],
   "source": [
    "print (\"Filtered PLPs files:\\n\")\n",
    "\n",
    "! ls -lah $output_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc14eb69",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e9970d49",
   "metadata": {},
   "source": [
    "# Numbers check for the paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5d93dba0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All PLPs in the cohort: 764216\n"
     ]
    }
   ],
   "source": [
    "filter_homozygous=True\n",
    "\n",
    "cohort_plp_files = glob.glob(f\"{output_folder}/chr*\")\n",
    "\n",
    "cohort_plps = VariantFeatures().read_sample_plps(cohort_plp_files, filter_homozygous=filter_homozygous).drop(['raw pos', 'raw ref', 'raw alt', 'alleles', 'GT.alleles'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "15e10a4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of PLPs in the cohort: 761211\n",
      "Number of unique PLPs: 54755\n"
     ]
    }
   ],
   "source": [
    "print (f\"Number of PLPs in the cohort:\", cohort_plps.shape[0])\n",
    "print (f\"Number of unique PLPs:\", cohort_plps[['chrom', 'pos', 'ref', 'alt']].drop_duplicates().shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4588bc66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of hom PLPs in the cohort: 0\n",
      "Number of unique hom PLPs: 0\n"
     ]
    }
   ],
   "source": [
    "homozygous = cohort_plps[cohort_plps['GT'] == '1/1']\n",
    "\n",
    "print (f\"Number of hom PLPs in the cohort:\", homozygous.shape[0])\n",
    "print (f\"Number of unique hom PLPs:\", homozygous[['chrom', 'pos', 'ref', 'alt']].drop_duplicates().shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bda05cc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>position</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>Hgvsc</th>\n",
       "      <th>Hgvsp</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>...</th>\n",
       "      <th>CADD_score</th>\n",
       "      <th>MOI-Pred_score</th>\n",
       "      <th>decipher</th>\n",
       "      <th>vkgl</th>\n",
       "      <th>hgmd-DM</th>\n",
       "      <th>clinvar</th>\n",
       "      <th>clinvar_stars</th>\n",
       "      <th>intervar</th>\n",
       "      <th>hets</th>\n",
       "      <th>homs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014051</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>EXON_REGION</td>\n",
       "      <td>False</td>\n",
       "      <td>c.47C&gt;A</td>\n",
       "      <td>p.Ser16Ter</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>...</td>\n",
       "      <td>29.400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>EXON_REGION</td>\n",
       "      <td>False</td>\n",
       "      <td>c.139C&gt;T</td>\n",
       "      <td>p.Gln47Ter</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>...</td>\n",
       "      <td>34.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014332</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>EXON_REGION</td>\n",
       "      <td>False</td>\n",
       "      <td>c.328C&gt;T</td>\n",
       "      <td>p.Gln110Ter</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>...</td>\n",
       "      <td>35.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>66</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014355</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>EXON_REGION</td>\n",
       "      <td>False</td>\n",
       "      <td>c.353del</td>\n",
       "      <td>p.Phe118SerfsTer20</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>...</td>\n",
       "      <td>27.100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014359</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>EXON_REGION</td>\n",
       "      <td>False</td>\n",
       "      <td>c.355G&gt;T</td>\n",
       "      <td>p.Glu119Ter</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>...</td>\n",
       "      <td>35.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54753</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50627390</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>ARSA</td>\n",
       "      <td>EXON_REGION</td>\n",
       "      <td>False</td>\n",
       "      <td>c.240dup</td>\n",
       "      <td>p.Gly81ArgfsTer53</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>...</td>\n",
       "      <td>30.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>Pathogenic/Likely_pathogenic</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54754</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50627584</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>ARSA</td>\n",
       "      <td>EXON_REGION</td>\n",
       "      <td>False</td>\n",
       "      <td>c.195del</td>\n",
       "      <td>p.Tyr65Ter</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>...</td>\n",
       "      <td>33.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>Pathogenic/Likely_pathogenic</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54755</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50627718</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>ARSA</td>\n",
       "      <td>EXON_REGION</td>\n",
       "      <td>False</td>\n",
       "      <td>c.61del</td>\n",
       "      <td>p.Arg21ValfsTer7</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>...</td>\n",
       "      <td>13.500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54756</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50627739</td>\n",
       "      <td>G</td>\n",
       "      <td>GC</td>\n",
       "      <td>ARSA</td>\n",
       "      <td>EXON_REGION</td>\n",
       "      <td>False</td>\n",
       "      <td>c.40dup</td>\n",
       "      <td>p.Ala14GlyfsTer62</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>...</td>\n",
       "      <td>12.040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54757</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50627757</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>ARSA</td>\n",
       "      <td>EXON_REGION</td>\n",
       "      <td>False</td>\n",
       "      <td>c.22del</td>\n",
       "      <td>p.Ser8ProfsTer20</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>...</td>\n",
       "      <td>1.688</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54758 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr  position ref alt   gene       region  synonymous     Hgvsc  \\\n",
       "0       chr1   1014051   C   A  ISG15  EXON_REGION       False   c.47C>A   \n",
       "1       chr1   1014143   C   T  ISG15  EXON_REGION       False  c.139C>T   \n",
       "2       chr1   1014332   C   T  ISG15  EXON_REGION       False  c.328C>T   \n",
       "3       chr1   1014355  CT   C  ISG15  EXON_REGION       False  c.353del   \n",
       "4       chr1   1014359   G   T  ISG15  EXON_REGION       False  c.355G>T   \n",
       "...      ...       ...  ..  ..    ...          ...         ...       ...   \n",
       "54753  chr22  50627390   C  CG   ARSA  EXON_REGION       False  c.240dup   \n",
       "54754  chr22  50627584  CG   C   ARSA  EXON_REGION       False  c.195del   \n",
       "54755  chr22  50627718  CG   C   ARSA  EXON_REGION       False   c.61del   \n",
       "54756  chr22  50627739   G  GC   ARSA  EXON_REGION       False   c.40dup   \n",
       "54757  chr22  50627757  GA   G   ARSA  EXON_REGION       False   c.22del   \n",
       "\n",
       "                    Hgvsp  variant_type  ... CADD_score  MOI-Pred_score  \\\n",
       "0              p.Ser16Ter  Substitution  ...     29.400             NaN   \n",
       "1              p.Gln47Ter  Substitution  ...     34.000             NaN   \n",
       "2             p.Gln110Ter  Substitution  ...     35.000             NaN   \n",
       "3      p.Phe118SerfsTer20      Deletion  ...     27.100             NaN   \n",
       "4             p.Glu119Ter  Substitution  ...     35.000             NaN   \n",
       "...                   ...           ...  ...        ...             ...   \n",
       "54753   p.Gly81ArgfsTer53     Insertion  ...     30.000             NaN   \n",
       "54754          p.Tyr65Ter      Deletion  ...     33.000             NaN   \n",
       "54755    p.Arg21ValfsTer7      Deletion  ...     13.500             NaN   \n",
       "54756   p.Ala14GlyfsTer62     Insertion  ...     12.040             NaN   \n",
       "54757    p.Ser8ProfsTer20      Deletion  ...      1.688             NaN   \n",
       "\n",
       "      decipher  vkgl  hgmd-DM                       clinvar clinvar_stars  \\\n",
       "0          NaN   NaN      NaN                           NaN           NaN   \n",
       "1          NaN   NaN        Y                    Pathogenic           0.0   \n",
       "2          NaN   NaN      NaN        Uncertain_significance           1.0   \n",
       "3          NaN   NaN      NaN                           NaN           NaN   \n",
       "4          NaN   NaN        Y             Likely_pathogenic           1.0   \n",
       "...        ...   ...      ...                           ...           ...   \n",
       "54753      NaN   NaN        Y  Pathogenic/Likely_pathogenic           2.0   \n",
       "54754      NaN   NaN        Y  Pathogenic/Likely_pathogenic           2.0   \n",
       "54755      NaN   NaN      NaN                           NaN           NaN   \n",
       "54756      NaN   NaN      NaN                           NaN           NaN   \n",
       "54757      NaN   NaN      NaN                           NaN           NaN   \n",
       "\n",
       "                intervar hets homs  \n",
       "0      Likely_pathogenic    1    0  \n",
       "1             Pathogenic    2    0  \n",
       "2             Pathogenic   66    0  \n",
       "3      Likely_pathogenic    1    0  \n",
       "4             Pathogenic   12    0  \n",
       "...                  ...  ...  ...  \n",
       "54753         Pathogenic   42    0  \n",
       "54754         Pathogenic    1    0  \n",
       "54755  Likely_pathogenic    2    0  \n",
       "54756  Likely_pathogenic    1    0  \n",
       "54757  Likely_pathogenic    1    0  \n",
       "\n",
       "[54758 rows x 24 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_plps = pd.read_csv(all_plps_file, sep='\\t')\n",
    "\n",
    "all_plps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18f1126e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ukbb_recessive",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
