{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import ukbb_recessive.regression.plotting as plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plps file\n",
    "plps = pd.read_csv(\n",
    "    \".../450k/plp_selection/basic/new_gene_names/\"\n",
    "    \"new_freq/new_relatedness/all_chr_total_presumable_plps_HFE_final_sorted.txt\", sep=\"\\t\")\n",
    "\n",
    "#gene panel\n",
    "gene_panel = pd.read_csv(\".../450k/regions/gene-panel-gencode-v34.txt\", header=None)\n",
    "gene_panel.columns = ['gene', 'panel']\n",
    "\n",
    "# samples of interest: European & non-related without hom and comp_het\n",
    "# european_non_rel_samples = (\n",
    "#     \".../450k/samples/european_non_related_no_withdrawal_to_include_450k.no_hom_comp_het.txt\"\n",
    "# )\n",
    "european_non_rel_samples = (\n",
    "    \".../450k/samples/european_non_related_no_withdrawal_to_include_450k.txt\"\n",
    ")\n",
    "\n",
    "with open(european_non_rel_samples, 'r') as f:\n",
    "    european_non_rel_samples = [l.strip() for l in f.readlines()]\n",
    "\n",
    "N = len(european_non_rel_samples)\n",
    "print (f\"Number of samples: {N}\")\n",
    "\n",
    "# mild plps; were taken from Hila's excel file; they are the mild variants\n",
    "mild_plps = pd.read_csv('mild_variants.csv', sep='\\t').rename(columns={'hets': 'mild_hets', 'homs': 'mild_homs'}).drop('panel', axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# separate mild variants\n",
    "plps = plps.merge(mild_plps, how='left')\n",
    "plps[['mild_hets', 'mild_homs']] = plps[['mild_hets', 'mild_homs']].fillna(0)\n",
    "plps['hets'] = plps['hets'] - plps['mild_hets']\n",
    "plps['homs'] = plps['homs'] - plps['mild_homs']\n",
    "\n",
    "# add gene panel information\n",
    "plps = plps.merge(gene_panel, how='inner', on='gene')\n",
    "plps = plps[~plps['panel'].isin(['Tumor', 'No_panel', 'Cardiovascular'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_data_per_gene(plps):\n",
    "    # per gene genotypes count\n",
    "    plps_gene = plps.groupby(['panel', 'gene'])[['hets', 'homs', 'mild_hets', 'mild_homs']].sum().reset_index()\n",
    "\n",
    "    # allele cound\n",
    "    plps_gene['AC'] = plps_gene['hets'] + 2*plps_gene['homs']\n",
    "    plps_gene['mild_AC'] = plps_gene['mild_hets'] + 2*plps_gene['mild_homs']\n",
    "\n",
    "    # allele frequency\n",
    "    plps_gene['AF'] = plps_gene['AC'] / (2*N)\n",
    "    plps_gene['mild_AF'] = plps_gene['mild_AC'] / (2*N)\n",
    "    plps_gene['total_AF'] = plps_gene['AF'] + plps_gene['mild_AF']\n",
    "\n",
    "    # heterozygotes probability\n",
    "    plps_gene['hets_expected'] = 2*plps_gene['AF']*(1-plps_gene['AF'])\n",
    "    plps_gene['mild_hets_expected'] = 2*plps_gene['mild_AF']*(1-plps_gene['mild_AF'])\n",
    "\n",
    "    # at-risk-couples\n",
    "    plps_gene['at_risk_couples'] = (plps_gene['hets_expected']**2 + 2*plps_gene['hets_expected']*plps_gene['mild_hets_expected']) \n",
    "    plps_gene['at_risk_couples_first_cousins'] = plps_gene['hets_expected']*0.125\n",
    "\n",
    "    plps_gene['CR'] = plps_gene['at_risk_couples_first_cousins'] / plps_gene['at_risk_couples'] \n",
    "\n",
    "    return plps_gene\n",
    "\n",
    "def add_data_per_panel(plps_gene):\n",
    "    # calculate CR on a panel level\n",
    "    plps_panel = plps_gene.groupby('panel').agg({'at_risk_couples': 'sum', 'at_risk_couples_first_cousins': 'sum', 'total_AF': 'std'}).reset_index()\n",
    "\n",
    "    plps_panel['affected_offsprings']  = plps_panel['at_risk_couples'] * 100000 / 4\n",
    "    plps_panel['affected_offsprings_first_cousins']  = plps_panel['at_risk_couples_first_cousins'] * 100000 / 4\n",
    "\n",
    "    plps_panel['CR'] = plps_panel['at_risk_couples_first_cousins'] / plps_panel['at_risk_couples']\n",
    "\n",
    "    return plps_panel\n",
    "\n",
    "plps_gene = add_data_per_gene(plps)\n",
    "plps_panel = add_data_per_panel(plps_gene)\n",
    "plps_panel = plps_panel.sort_values('CR', ascending=False).reset_index(drop=True)\n",
    "\n",
    "plps_panel.to_csv('../../../../data/tables/CR.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "\n",
    "for threshold in range(1, 5350, 10):\n",
    "    plps_rare = plps[(plps['hets'] <= threshold)].copy()\n",
    "    plps_gene = add_data_per_gene(plps_rare)\n",
    "    plps_panel = add_data_per_panel(plps_gene)\n",
    "\n",
    "    plps_panel = plps_panel.sort_values('CR', ascending=False).reset_index(drop=True)\n",
    "    plps_panel['rank'] = plps_panel.index + 1\n",
    "    plps_panel['threshold'] = threshold\n",
    "    results.append(plps_panel)\n",
    "\n",
    "results = pd.concat(results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sns.scatterplot(data=results[results['panel'].isin(['ID-total', 'Neuromuscular'])], x='threshold', y='rank', hue='panel')\n",
    "\n",
    "sns.scatterplot(data=results[results['panel'].isin(['ID-total'])], x='threshold', y='rank', hue='panel')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(data=results[results['panel'].isin([ 'Neuromuscular'])], x='threshold', y='rank', hue='panel')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results[results['threshold'] == 1051]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ukbb_recessive",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
